hd44780: Add support for libgpiod2

manio opened the pull request:
Hi,
The old libgpiod1x library is nowadays not available in new OS releases (it can't even be compiled on my old `raspbian`), so change the original @haraldg gpiod1x driver implementation to add a support for the newest libgpiod2 library and decent kernel implementations.

This is fixing #214 and #217.

If you're OK with the changes, please do normal merge and don't squash because I've split the changes into separate commits for easy reviewing in the future.

manio committed:
hd44780/gpiod: rewrite to support libgpiod2

Closes #214, #217.

manio commented:
Just a side note: I was using my raspberry-connected LCD display using:
`ConnectionType=raspberrypi`
and it was working fine just after start, but then later it always ended up with a garbage (I was trying with `RefreshDisplay=`, `DelayMult=`, etc. but doesn't help).

Now when using my new libgpiod2 implementation, it is working like a charm! :)

haraldg commented:
Thanks for looking into this!

First question: Does this code work with the old and new version of libgpiod or only the new one?

Rant: I'm quite annoyed, that libgpiod broke the API without a way to have both versions coexist (like libusb did) and even without a standardized (future proof) way to check the installed version. What a mess.

> please do normal merge and don't squash because I've split the changes into separate commits for easy reviewing in the future.

Actually your way to split the changes into commits makes it much harder to review them, because they are not independent. After spending about 30 Minutes switching back and forth between commits and cross-referencing things, I gave up.

As a guideline every commit should be useful on its own. At the very least the code should compile. Actually every commit should be compile tested.

Things like renaming arguments without changing the function body are a no go.

Please resubmit.

manio commented:
> Thanks for looking into this!
> 
> First question: Does this code work with the old and new version of libgpiod or only the new one?

only the new one

> Rant: I'm quite annoyed, that libgpiod broke the API without a way to have both versions coexist (like libusb did) and even without a standardized (future proof) way to check the installed version. What a mess.
> 
> > please do normal merge and don't squash because I've split the changes into separate commits for easy reviewing in the future.
> 
> Actually your way to split the changes into commits makes it much harder to review them, because they are not independent. After spending about 30 Minutes switching back and forth between commits and cross-referencing things, I gave up.

you could just go to https://github.com/lcdproc/lcdproc/pull/219/files and you'd have all changes on one screen, and if you do it locally you could do `git diff` with range. Or just a `git diff` on a whole merge commit.

> As a guideline every commit should be useful on its own. At the very least the code should compile. Actually every commit should be compile tested.

I spent a lot of time committing the changes as logical steps instead of one big "blob". I know it is not compiling on every step but this is just what merge commits are for. This is just a way I am doing my work, but OK - no problem - I am adjusting to your approach: I've squashed this for you and forced-pushed.

> Things like renaming arguments without changing the function body are a no go.

Sorry - what you mean by this? Can you do an inline-comment review?

haraldg commented:
> > Things like renaming arguments without changing the function body are a no go.
> 
> Sorry - what you mean by this? Can you do an inline-comment review?

In your old commit 0afa686c67187005ab0ad9571e63cb4d32e3bce2 you changed the argument to `init_gpio_pin()` from `pin` to `gpin` but you didn't make the same change in the body of the function. (You did it in a later commit.)

Your idea of logical commits is sound. I like to work like that myself and demand it, when reviewing bigger changes. However you didn't split at the right places.

If I have to use `git diff` on a range of commits during review, then clearly splitting wasn't helpful in the first place, was it?

haraldg commented on server/drivers/hd44780-gpiod.c:
```diff
@@ -35,19 +35,41 @@ void gpiod_HD44780_backlight(PrivateData *p, unsigned char state);
+ * \param value   The new PIN state to set.
+ */
+static int
+gpiod_line_set_value(gpio_pin *pin, int value)
```
Adding a local function with the same name as an (old) API function (or even just in the libgpiod "namespace") is confusing. I can see us do this, if you do a version check and write code, that supports both versions.

But otherwise I'd rename this function. Personally I'd just drop it and use the proper calls directly. They are not that complex.

manio replied:
I'll rather prefer to choose option to rename that function, otherwise we would end up with calls like:
`gpiod_line_request_set_value(pin->req, pin->offset, (state == BACKLIGHT_ON) ? GPIOD_LINE_VALUE_ACTIVE : GPIOD_LINE_VALUE_INACTIVE);`

haraldg commented on server/drivers/hd44780-gpiod.c:
```diff
@@ -58,29 +80,39 @@ typedef struct {
-		return -1;
-	}
+	*gpin = calloc(1, sizeof(gpio_pin));
+	gpio_pin *pin = *gpin;
```
IIRC in body declarations are not LCDproc coding style.

Anyway, the naming convention between `pin` and `gpin`seems somewhat unintuitive. Probably the easiest to understand solution would be, to just add the asterisk each time to de-reference the double pointer.

If instead you want to change the function to return a pointer directly instead of passing the double pointer, that would also be okay.

(Yes, nitpicking that one. I wouldn't bother, if other changes weren't needed.)

haraldg commented on server/drivers/hd44780-gpiod.c:
```diff
@@ -90,6 +122,28 @@ init_gpio_pin(Driver *drvthis, struct gpiod_line **pin, const char *name)
+ * \param pin     Pointer to gpio_pin structure.
+ */
+void
+release_gpio_pin(gpio_pin *pin)
```
declaration as `static` missing

haraldg commented on server/drivers/hd44780-gpiod.c:
```diff
@@ -90,6 +122,28 @@ init_gpio_pin(Driver *drvthis, struct gpiod_line **pin, const char *name)
+		gpiod_line_settings_free(pin->settings);
+	if (pin->req)
+		gpiod_line_request_release(pin->req);
+}
```
I think, you forgot to free `pin` here.

haraldg commented on server/drivers/hd44780-gpiod.c:
```diff
@@ -145,9 +200,10 @@ hd_init_gpiod(Driver *drvthis)
 
-	pins->chip = gpiod_chip_open_lookup(chip);
+	snprintf(dev_path, sizeof(dev_path), "/dev/gpiochip%s", chip);
+	pins->chip = gpiod_chip_open(dev_path);
```
I think, this changes behavior. But it wasn't explicitly documented behavior, so I hope it is okay.

It's a bit unfortunate though. The old call allowed to pass some unique identifier like `gpio.085f3b00`, that at least some systems have. If we are now back to sequentially indexed devices, then unstable probing order at boot time might actually break things.

manio replied:
Yes, I was also thinking how to handle it when I was writing that code. But as you said it is currently undocumented, so I've changed the approach and I think the most flexible for user is to let this be a full path of a gpio device. User may also want to use eg. some UDEV rules (something similar like `/dev/disk/by-id...` - so restricting to `/dev/SOMETHING` is maybe not a good choice).

manio commented:
OK... pushed, all your code requests addressed.

haraldg commented:
I just added some minor updates to the documentation and committed this PR in https://github.com/lcdproc/lcdproc/commit/71877ee059a238400b5f6b41cee7c43a3df00334

manio commented:
Thank you!
